modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -0,0 +1,258 @@
+        updateTaskState(name);
+    }
+
+    boolean updateTaskState(String name) {
+        PlainActionFuture<PersistentTask<?>> future = PlainActionFuture.newFuture();
+        persistentTask.updatePersistentTaskState(state, future);
+        try {
+            state = ((GeoIpTaskState) future.actionGet().getState()).copy();
+            return true;
+        } catch (Exception e) {
+            logger.error("can not update state during geoip database [" + name + "] update", e);
+            return false;
+        }
+    }
+
+    //visible for testing
+    int indexChunks(String name, InputStream is, int chunk) throws IOException {
+        while (true) {
+            byte[] buf = getChunk(is);
+            if (buf.length == 0) {
+                break;
+            }
+            client.prepareIndex(DATABASES_INDEX).setId(name + "_" + chunk)
+                .setSource(XContentType.SMILE, "name", name, "chunk", chunk, "data", buf)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setWaitForActiveShards(ActiveShardCount.ALL)
+                .get();
+            chunk++;
+        }
+        return chunk;
+    }
+
+    //visible for testing
+    byte[] getChunk(InputStream is) throws IOException {
+        byte[] buf = new byte[MAX_CHUNK_SIZE];
+        int chunkSize = 0;
+        while (chunkSize < MAX_CHUNK_SIZE) {
+            int read = is.read(buf, chunkSize, MAX_CHUNK_SIZE - chunkSize);
+            if (read == -1) {
+                break;
+            }
+            chunkSize += read;
+        }
+        if (chunkSize < MAX_CHUNK_SIZE) {
+            buf = Arrays.copyOf(buf, chunkSize);
+        }
+        return buf;
+    }
+
+    //for testing only
+    void setPersistentTask(AllocatedPersistentTask persistentTask) {
+        this.persistentTask = persistentTask;
+    }
+
+    //for testing only
+    void setState(GeoIpTaskState state) {
+        this.state = state;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask allocatedTask, PersistentTaskParams params, PersistentTaskState state) {
+        if (scheduled != null) {
+            scheduled.cancel();
+        }
+        if (enabled) {
+            this.persistentTask = allocatedTask;
+            this.state = state != null ? (GeoIpTaskState) state : new GeoIpTaskState();
+            try {
+                updateDatabases();
+            } catch (Exception e) {
+                logger.error("exception during geoip databases update", e);
+            }
+            scheduleNextRun(pollInterval);
+        }
+    }
+
+    private void scheduleNextRun(TimeValue time) {
+        scheduled = threadPool.schedule(() -> nodeOperation(persistentTask, null, state), time, ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            clusterService.removeListener(this);
```

Comment:
maybe add a comment that this is needed to bootstrap the downloader?
(after this, bootstrapping isn't needed, because in case of failure persistent tasks get started automatically on other nodes)

Reply:
I've tried it but I couldn't make it work with out security manager + it leaves threads behind (without obvious way how to close/clean it) which our tests don't like. I'd leave it as it is for now if you don't mind.

Reply:
Was this reply meant for the comment about http client?

Reply:
Yes, of course, my bad - clicked the wrong text box